lib/pos_staking_rewards.go

```diff
@@ -0,0 +1,282 @@
+) (
+	_stakerRewardNanos uint64,
+	_validatorCommissionNanos uint64,
+	_err error,
+) {
+	// Compute the staker's reward amount using a big float math, and immediately convert it to big int
+	// so we can do the remainder of the math using integer operations. This is the only operation where
+	// we need float math.
+	stakerRewardNanos := convertBigFloatToBigInt(
+		computeStakingReward(snapshotStakeEntry.StakeAmountNanos, elapsedFractionOfYear, apy),
+	)
+
+	// If the reward is 0, then there's nothing to be done. In practice, the reward should never be < 0
+	// either, but we check for it here in case it resulted from a rounding error. Either way, we're
+	// safe to exit early here.
+	if stakerRewardNanos == nil || stakerRewardNanos.Sign() <= 0 {
+		return 0, 0, nil
+	}
+
+	// At this point, we know that the staker has non-zero rewards. We need to determine how to
+	// distribute the rewards to them, and how to distribute the validator's commissions.
+
+	// Compute the validator's commission and deduct it from the staker's reward.
+	validatorCommissionNanos := big.NewInt(0)
+
+	// We only compute validator commission if the staker had delegated stake to another validator. If the staker
+	// staked to themselves, then there's no reason to compute the validator commission.
+	if !snapshotStakeEntry.StakerPKID.Eq(snapshotStakeEntry.ValidatorPKID) {
+		// Fetch the ValidatorEntry that the stake is delegated to. The validator is guaranteed to be in the
+		// snapshot validator set, because only stakes from the snapshot validator set are eligible to receive
+		// rewards.
+		validatorEntry, err := bav.GetSnapshotValidatorSetEntryByPKID(snapshotStakeEntry.ValidatorPKID)
+		if err != nil {
+			return 0, 0, errors.Wrapf(err, "computeStakerRewardAndValidatorCommission: problem fetching validator entry: ")
+		}
+		if validatorEntry == nil {
+			// This should never happen. If we can't find the validator, then something is wrong. It's safest to error
+			// and return early here.
+			return 0, 0, errors.Errorf("computeStakerRewardAndValidatorCommission: validator entry should never be nil")
+		}
+
+		if validatorEntry.DelegatedStakeCommissionBasisPoints > 0 {
+			// We use integer math to compute the validator's commission. The commission is computed as:
+			// floor(stakerReward * validatorCommissionBasisPoints / 10000)
+			validatorCommissionNanos = computeValidatorCommission(
+				stakerRewardNanos, validatorEntry.DelegatedStakeCommissionBasisPoints,
+			)
+
+			if validatorCommissionNanos.Cmp(stakerRewardNanos) > 0 {
+				// This should never happen. If the validator's commission is greater than the total staker reward amount,
+				// then something has gone wrong.
+				return 0, 0, errors.Errorf(
+					"computeStakerRewardAndValidatorCommission: validator commission is greater than staker reward amount",
+				)
+			}
+
+			// Deduct the validator commission from the staker's reward.
+			stakerRewardNanos.Sub(stakerRewardNanos, validatorCommissionNanos)
```

Comment:
Prefer `stakerRewardNanos = big.NewInt(0).Sub(stakerRewardNanos, validatorCommissionNanos)`

The reason we prefer this format is it's less error-prone. Technically, it's more efficient to modify a pointer in-place like you have it here, but it's harder to read and runs the risk of causing bugs.